File: vscodeExt/202.happy-number.py
#
# @lc app=leetcode id=202 lang=python3
#
# [202] Happy Number
#

# @lc code=start
class Solution:
    def isHappy(self, n: int) -> bool:
        # 记录过程，防止产生循环。
        s = set()
        while n not in s:
            s.add(n)
            sum = 0 
            while n!= 0:
                sum += (n%10) ** 2 
                n = n//10
            print(sum)    
            if sum == 1:
                return True
            n = sum         
        # 构成循环，返回false 
        else:
            return False
# @lc code=end

File: vscodeExt/202.happy-number.java
import java.util.HashSet;
import java.util.Set;

/*
 * @lc app=leetcode id=202 lang=java
 *
 * [202] Happy Number
 */

// @lc code=start
class Solution {
    public boolean isHappy(int n) {

        Set<Integer> s = new HashSet<>();
        while(!s.contains(n)){
            // System.out.println(n);
            s.add(n);
            n = squareSum(n);
            if(n == 1){
                return true;
            } 
        }

        return false;
    }

    public int squareSum(int n){
        int res = 0;
        while(n!=0){
            res += (n%10) * (n%10);
            n /= 10;
        }
        return res;
    }
}
// @lc code=end

